hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -1028,6 +946,122 @@ func (r *NodePoolReconciler) reconcile(ctx context.Context, hcluster *hyperv1.Ho
+			Status:             corev1.ConditionUnknown,
+			Reason:             hyperv1.NodePoolFailedToGetReason,
+			Message:            err.Error(),
+			ObservedGeneration: nodePool.Generation,
+		})
+		return fmt.Errorf("failed to get Machines: %w", err)
+	}
+
+	r.setAllMachinesReadyCondition(nodePool, machines)
+
+	r.setAllNodesHealthyCondition(nodePool, machines)
+
+	return nil
+}
+
+func (r *NodePoolReconciler) setAllNodesHealthyCondition(nodePool *hyperv1.NodePool, machines []*capiv1.Machine) {
+	status := corev1.ConditionTrue
+	reason := hyperv1.AsExpectedReason
+	var message string
+
+	if len(machines) < 1 {
+		status = corev1.ConditionFalse
+		reason = hyperv1.NodePoolNotFoundReason
+		message = "No Machines are created"
+	}
+
+	for _, machine := range machines {
+		condition := findCAPIStatusCondition(machine.Status.Conditions, capiv1.MachineNodeHealthyCondition)
+		if condition != nil && condition.Status != corev1.ConditionTrue {
+			status = corev1.ConditionFalse
+			reason = condition.Reason
+			message = message + fmt.Sprintf("Machine %s: %s\n", machine.Name, condition.Reason)
+		}
+	}
+
+	if status == corev1.ConditionTrue {
+		message = hyperv1.AllIsWellMessage
+	}
+
+	allMachinesHealthyCondition := &hyperv1.NodePoolCondition{
+		Type:               hyperv1.NodePoolAllNodesHealthyConditionType,
+		Status:             status,
+		Reason:             reason,
+		Message:            message,
+		ObservedGeneration: nodePool.Generation,
+	}
+	SetStatusCondition(&nodePool.Status.Conditions, *allMachinesHealthyCondition)
+}
+
+func (r *NodePoolReconciler) setAllMachinesReadyCondition(nodePool *hyperv1.NodePool, machines []*capiv1.Machine) {
+	status := corev1.ConditionTrue
+	reason := hyperv1.AsExpectedReason
+	var message string
+
+	if len(machines) < 1 {
+		status = corev1.ConditionFalse
+		reason = hyperv1.NodePoolNotFoundReason
+		message = "No Machines are created"
+	} else {
+		// Aggregate conditions.
+		// TODO (alberto): consider bubbling failureReason / failureMessage.
+		// This a rudimentary approach which aggregates every Machine, until
+		// https://github.com/kubernetes-sigs/cluster-api/pull/6218 and
+		// https://github.com/kubernetes-sigs/cluster-api/pull/6025
+		// are solved.
+		// Eventually we should solve this in CAPI to make it available in MachineDeployments / MachineSets
+		// with a consumable "Reason" and an aggregated "Message".
+
+		var messages []string
+		for _, machine := range machines {
+			readyCond := findCAPIStatusCondition(machine.Status.Conditions, capiv1.ReadyCondition)
+			if readyCond != nil && readyCond.Status != corev1.ConditionTrue {
+				status = corev1.ConditionFalse
+				infraReadyCond := findCAPIStatusCondition(machine.Status.Conditions, capiv1.InfrastructureReadyCondition)
+				if infraReadyCond != nil && infraReadyCond.Status != corev1.ConditionTrue && !isSetupCounterCondMessage.MatchString(infraReadyCond.Message) {
+					reason = infraReadyCond.Reason
```

Comment:
I'm unsure about how `reason` is handled here. It's possible that there are multiple reasons for an infrastructure issue within a nodepool. For example, one machine might get stuck scheduling due to resource limitations and another might be failing due to a crash or infra node failure.

The capi project has some sophisticated logic for merging conditions and bubbling up the reason with the highest severity. I'm not sure I agree with that logic because it potentially obscures a full view of what is happening. 

Maybe we could consider merging all the reasons together into one string and removing duplicate entries. 

thoughts?